File: battleship/logic.py
"""
This module is describing high-level control flow of the game, that is
essentially logic behind "game rules".
This is typical "controller" in MVC terminology.
"""
import board
import fsm

BATTLE_STARTED='BATTLE_STARTED'; PLAYER_TURN='PLAYER_TURN'; BATTLE_ENDED='BATTLE_ENDED' # states
SETUP='SETUP'; SHOOT='SHOOT'; RESTART='RESTART' # events
SEA_SIDE = 10

# A "facade" around particular gamelogic implementation
class Game:
	"""
	Usage like this:

	game = logic.Game("player one", "player two")
	while True:
		state = game.get_state()
		if state == logic.BATTLE_STARTED:
			ships_for_player_one = ...
			game.setup("player one", ships_for_player_one)
			ships_for_player_two = ...
			game.setup("player two", ships_for_player_two)
		elif state == logic.PLAYER_TURN:
			current_player_id = game.current_player()
			the_shot = ...
			shot_result = game.shot(current_player_id, the_shot)
		elif state == logic.BATTLE_ENDED:
			break
	"""
	config = (
		(1, 4),  # one battleship
		(2, 3),  # two cruisers
		(3, 2),  # three destroyers
		(4, 1))  # four boats

	def __init__(self, player1, player2):
		self.context = {
			'players': {
				player1: {'ready': False, 'opponent': player2},
				player2: {'ready': False, 'opponent': player1}},
			'current': None}
		self.state_machine = fsm.Transducer(states, BATTLE_STARTED, self.context)

	def get_state(self):
		return self.state_machine.get_state()

	def setup(self, player, ships):
		self.state_machine.dispatch(SETUP, (player, ships))

	def shoot(self, player, shot):
		assert self.context['current'] == player
		self.state_machine.dispatch(SHOOT, (player, shot))
		return self.context['players'][player]['last_shot']

	def current_player(self):
		return self.context['current']

	def opponent_ships(self, player):
		opponent = self.context['players'][player]['opponent']
		return self.context['players'][opponent]['sea'].ships

	def player_shots(self, player):
		opponent = self.context['players'][player]['opponent']
		return self.context['players'][opponent]['sea'].shots

# Rule defenitions and support

def restart_battle(context):  # entry action
	for player in context['players'].itervalues():
		player['sea'] = None
	context['current'] = context['players'].keys()[0]

def does_all_players_set(context):  # condition
	for player in context['players'].itervalues():
		if not player['ready']:
			return False
	return True

def is_shot_missed(context):  # condition
	current_player = context['players'][context['current']]
	return current_player['last_shot'] == 'miss'

def does_opponent_has_no_ships(context):  # condition
	current_player = context['players'][context['current']]
	target_sea = context['players'][current_player['opponent']]['sea']
	if len(target_sea.active_ships) == 0:
		return True
	else:
		return False

def pass_turn(context):  # action
	current_player = context['players'][context['current']]
	context['current'] = current_player['opponent']

def setup_ships(context, input):  # event input
	player, ships = input
	context['players'][player]['sea'] = board.SeaGrid(SEA_SIDE, ships)
	context['players'][player]['ready'] = True

def player_shoot(context, input):  # event input
	player, pos = input
	current_player = context['players'][context['current']]
	target_sea = context['players'][current_player['opponent']]['sea']
	current_player['last_shot'] = target_sea.shoot_square(pos)

states = {
	BATTLE_STARTED: fsm.State(
		on_enter = restart_battle,
		events = {
			SETUP: fsm.Event(
				input = setup_ships,
				transitions = (
					fsm.Transition(condition=does_all_players_set, state=PLAYER_TURN),))}),
	PLAYER_TURN: fsm.State(
		events = {
			SHOOT: fsm.Event(
				input = player_shoot,
				transitions = (
					fsm.Transition(condition=is_shot_missed, state=PLAYER_TURN, action=pass_turn),
					fsm.Transition(condition=does_opponent_has_no_ships, state=BATTLE_ENDED)))}),
	BATTLE_ENDED: fsm.State(
		events = {
			RESTART: fsm.Event(
				transitions = (
					fsm.Transition(state=BATTLE_STARTED),))})}

File: battleship/game/logic.cpp
#include "logic.h"
#include "fsm.h"

#include <assert.h>
#include <map>


namespace logic {


struct PlayerState
{
	bool ready;
	PLAYER_HANDLE opponent;
	board::SHOT last_shot;
	board::SeaGrid* sea;
	PlayerState() // DELME
	{
		opponent = 0;
		ready = false;
		sea = NULL;
	}
	PlayerState(PLAYER_HANDLE opponent_handle)
	{
		opponent = opponent_handle;
		ready = false;
		sea = NULL;
	}
};

struct GameContext
{
	PLAYER_HANDLE current;
	std::map<PLAYER_HANDLE, PlayerState> players;
	GameContext(PLAYER_HANDLE player1, PLAYER_HANDLE player2)
	{
		players[player1] = PlayerState(player2);
		players[player2] = PlayerState(player1);
	}
	~GameContext()
	{
		for (std::map<PLAYER_HANDLE, PlayerState>::iterator it=players.begin(); it != players.end(); it++)
		{
			if (it->second.sea != NULL)
			{
				delete it->second.sea;
				it->second.sea = NULL;
			}
		}
	}
};

enum EVENTDATATYPE { SHIPS, POS };

struct EventData
{
	PLAYER_HANDLE player;
	EVENTDATATYPE type;
	void* data;
	EventData(PLAYER_HANDLE p, const std::vector<board::ShipAnchor>* ships)
	{
		player = p;
		type = SHIPS;
		data = const_cast<std::vector<board::ShipAnchor>*>(ships);
	}
	EventData(PLAYER_HANDLE p, const board::Pos* pos)
	{
		player = p;
		type = POS;
		data = const_cast<board::Pos*>(pos);
	}
};


// Rule defenitions and support

void restart_battle(GameContext* context)  // entry action
{
	for (std::map<PLAYER_HANDLE,PlayerState>::iterator it = context->players.begin();
		it != context->players.end();
		it++)
	{
		PlayerState& player = it->second;
		player.sea = NULL;
	}
	context->current = context->players.begin()->first;
}

bool does_all_players_set(GameContext* context)  // condition
{
	for (std::map<PLAYER_HANDLE,PlayerState>::iterator it = context->players.begin();
		it != context->players.end();
		it++)
	{
		PlayerState& player = it->second;
		if (!player.ready)
			return false;
	}
	return true;
}

bool is_shot_missed(GameContext* context)  // condition
{
	PlayerState& current_player = context->players[context->current];
	return current_player.last_shot == board::MISS;
}

bool does_opponent_has_no_ships(GameContext* context)  // condition
{
	PlayerState& current_player = context->players[context->current];
	board::SeaGrid* target_sea = context->players[current_player.opponent].sea;
	return target_sea->active_ships == 0;
}

void pass_turn(GameContext* context)  // action
{
	context->current = context->players[context->current].opponent;
}

bool setup_ships(GameContext* context, const EventData& input)  // event input
{
	assert (input.type == SHIPS);
	const std::vector<board::ShipAnchor>& ships = *reinterpret_cast<std::vector<board::ShipAnchor>*>(input.data);
	if (context->players[input.player].sea != NULL)
		delete context->players[input.player].sea;
	context->players[input.player].sea = new board::SeaGrid(10, ships);
	context->players[input.player].ready = true;
	return true;
}

bool player_shoot(GameContext* context, const EventData& input)  // event input
{
	assert (input.player == context->current);
	assert (input.type == POS);
	board::Pos& pos = *reinterpret_cast<board::Pos*>(input.data);
	PlayerState& current_player = context->players[context->current];
	board::SeaGrid* target_sea = context->players[current_player.opponent].sea;
	current_player.last_shot = target_sea->ShootSquare(pos);
	return true;
}

Game::Game(PLAYER_HANDLE player1, PLAYER_HANDLE player2)
{
	m_config.push_back(FleetConf(1, 4));  // one battleship
	m_config.push_back(FleetConf(2, 3));  // two cruisers
	m_config.push_back(FleetConf(3, 2));  // three destroyers
	m_config.push_back(FleetConf(4, 1));  // four boats
	m_context = new GameContext(player1, player2);
	std::map<int, fsm::State<GameContext, EventData> > states;

	states[BATTLE_STARTED] = fsm::State<GameContext,EventData>();
	states[BATTLE_STARTED].onEnter = restart_battle;
	states[BATTLE_STARTED].events[SETUP] = fsm::Event<GameContext,EventData>();
	states[BATTLE_STARTED].events[SETUP].input = setup_ships;
		fsm::Transition<GameContext,EventData> tr;
		tr.condition = does_all_players_set;
		tr.state = PLAYER_TURN;
	states[BATTLE_STARTED].events[SETUP].transitions.push_back(tr);

	states[PLAYER_TURN] = fsm::State<GameContext,EventData>();
	states[PLAYER_TURN].events[SHOOT] = fsm::Event<GameContext,EventData>();
	states[PLAYER_TURN].events[SHOOT].input = player_shoot;
		fsm::Transition<GameContext,EventData> player_turn_tr;
		player_turn_tr.state = PLAYER_TURN;
		player_turn_tr.condition = is_shot_missed;
		player_turn_tr.action = pass_turn;
	states[PLAYER_TURN].events[SHOOT].transitions.push_back(player_turn_tr);
		fsm::Transition<GameContext,EventData> battle_ended_tr;
		battle_ended_tr.state = BATTLE_ENDED;
		battle_ended_tr.condition = does_opponent_has_no_ships;
	states[PLAYER_TURN].events[SHOOT].transitions.push_back(battle_ended_tr);

	states[BATTLE_ENDED] = fsm::State<GameContext,EventData>();
	states[BATTLE_ENDED].events[RESTART] = fsm::Event<GameContext,EventData>();
	fsm::Transition<GameContext,EventData> battle_started_tr;
	battle_started_tr.state = BATTLE_STARTED;
	states[BATTLE_ENDED].events[RESTART].transitions.push_back(battle_started_tr);

	m_fsm = new fsm::Transducer<GameContext, EventData>(states, BATTLE_STARTED, m_context);
}
Game::~Game()
{
	delete m_context;
	delete m_fsm;
}

STATE Game::GetState()
{
	return (STATE)m_fsm->GetState();
}

void Game::Setup(PLAYER_HANDLE player, const std::vector<board::ShipAnchor>& ships)
{
	EventData data(player, &ships);  // (player, ships)
	m_fsm->Dispatch(SETUP, data);
}

board::SHOT Game::Shoot(PLAYER_HANDLE player, const board::Pos& shot)
{
	assert (m_context->current == player);
	EventData data(player, &shot);
	m_fsm->Dispatch(SHOOT, data);
	return m_context->players[player].last_shot;
}

PLAYER_HANDLE Game::GetCurrentPlayer()
{
	return m_context->current;
}

const std::vector<board::Ship> Game::GetOpponentShips(PLAYER_HANDLE player)
{
	PLAYER_HANDLE opponent = m_context->players[player].opponent;
	return m_context->players[opponent].sea->ships;
}

const std::vector<board::Pos> Game::GetPlayerShots(PLAYER_HANDLE player)
{
	PLAYER_HANDLE opponent = m_context->players[player].opponent;
	return m_context->players[opponent].sea->shots;
}


}  // namespace
